07-folds-monoids/JoinList.hs

```diff
@@ -0,0 +1,244 @@
+that yields a new JoinList whose monoidal annotation is derived from those of the two arguments.
+(+++) :: Monoid m => JoinList m a -> JoinList m a -> JoinList m a
+You may find it helpful to implement a helper function
+tag :: Monoid m => JoinList m a -> m
+which gets the annotation at the root of a JoinList
+-}
+--Exercise 1
+(+++) :: Monoid m => JoinList m a -> JoinList m a -> JoinList m a
+(+++) jl1 jl2 = Append (tag jl1 <> tag jl2) jl1 jl2
+    
+
+
+{-
+Exercise 2 The first annotation to try out is one for fast indexing into a JoinList. The idea is to cache the size (number of data ele- ments) of each subtree. This can then be used at each step to deter- mine if the desired index is in the left or the right branch.
+We have provided the Sized module that defines the Size type, which is simply a newtype wrapper around an Int. In order to make Sizes more accessible, we have also defined the Sized type class which provides a method for obtaining a Size from a value.
+Use the Sized type class to write the following functions.
+-}
+--Exercise 2
+{-
+Implement the function
+indexJ :: (Sized b, Monoid b) =>
+Int -> JoinList b a -> Maybe a
+indexJ finds the JoinList element at the specified index. If the index is out of bounds, the function returns Nothing. By an index in a JoinList we mean the index in the list that it represents. That is, consider a safe list indexing function
+(!!?) :: [a] -> Int -> Maybe a [] !!? _ = Nothing
+!!? i | i < 0 = Nothing (x:xs) !!? 0 =Justx (x:xs) !!? i = xs !!? (i-1)
+which returns Just the ith element in a list (starting at zero) if such an element exists, or Nothing otherwise. We also consider an updated function for converting join-lists into lists, just like jlbToList but ignoring the monoidal annotations:
+jlToList :: JoinList m a -> [a]
+jlToList Empty = []
+jlToList (Single _ a) = [a]
+jlToList (Append _ l1 l2) = jlToList l1 ++ jlToList l2
+We can now specify the desired behavior of indexJ. For any index i and join-list jl, it should be the case that
+   (indexJ i jl) == (jlToList jl !!? i)
+That is, calling indexJ on a join-list is the same as first convert- ing the join-list to a list and then indexing into the list. The point, of course, is that indexJ can be more efficient (O(log n) versus O(n), assuming a balanced join-list), because it gets to use the size annotations to throw away whole parts of the tree at once, whereas the list indexing operation has to walk over every element.
+-}
+indexJ :: (Sized b, Monoid b) => Int -> JoinList b a -> Maybe a
+indexJ n (Single _ a)
+    | n == 0    = Just a
+    | otherwise = Nothing
+indexJ n (Append m jl1 jl2)    
+    | n < jl1Size = indexJ n             jl1
+    | otherwise   = indexJ (n - jl1Size) jl2
+      where
+        jl1Size :: Int
+        jl1Size = getSize . size . tag $ jl1
+indexJ _ _ = Nothing
```

Comment:
Seems to have an unnecessary guard clause

```suggestion
indexJ 0 (Single _ a) = Just a
indexJ n (Append m jl1 jl2)    
    | n < jl1Size = indexJ n             jl1
    | otherwise   = indexJ (n - jl1Size) jl2
      where
        jl1Size :: Int
        jl1Size = getSize . size . tag $ jl1
indexJ _ _ = Nothing
```